Unlock platinum trophies

Megamouse opened the pull request:
This should unlock platinum trophies once you unlock the last necessary regular trophy.
Please back up your save and trophy dirs before testing.

isshininu commented:
Holy shit

https://github.com/RPCS3/rpcs3/issues/4289 this will close it

CookiePLMonster commented on rpcs3/Loader/TROPUSR.cpp:
```diff
@@ -215,6 +216,46 @@ u32 TROPUSRLoader::GetUnlockedTrophiesCount()
+				}
+
+				// Onto the next trophy id
+				break;
```
Unless I'm misunderstanding the intent, this begs for a `std::find_if`,

Megamouse replied:
Yeah I was just too lazy because I had no idea if this would work anyway

Megamouse commented:
> Holy shit
> 
> #4289 this will close it
> [RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/3964289/RPCS3.log.gz)
> Edit: not working on DeS platinum

Can you send me the trophy file that comes with the game?

isshininu commented:
This?
[TROPDIR.zip](https://github.com/RPCS3/rpcs3/files/3964314/TROPDIR.zip)

Megamouse commented:
I probably have to unlock the trophy myself instead of only telling the game that the requirements were met. Needs some more investigation.

Megamouse commented:
It should work now

isshininu commented:
On game load, nothing changes. Unfortunately I forgot to make a backup with 2 achievements left, so here is only platinum remains unlocked.

Megamouse commented:
I guess we'll never know since you seem to be the only tester

Megamouse commented:
In case you still have the savegame I can revert your trophy states. Maybe evenone that's easier to achieve

isshininu commented:
What exactly you want me to do?
If you want to edit out one of the achievements, any of the full upgrade weapon trophies are looks most easy to revert, as I can try to add it immediately again.

Megamouse commented:
The one in home/trophystuff/gameid

isshininu commented:
[NPWR01249_00.zip](https://github.com/RPCS3/rpcs3/files/3975389/NPWR01249_00.zip)
Sure, here it is.

Megamouse commented:
@isshininu I reverted the last change because we have some possibly related changes in master now.
I added some hacky code that lets you lock/unlock trophies by double click in the trophy manager while the game is stopped.

Could you please backup your savegames and trophies and try to unlock one of the last trophies (**not premium**)?

If this doesn't work then I will add my newer code again and we can hopefully make this work.

isshininu commented:
It looks like locking trophy in manager resets right after closing trophy manager. When opening it again, all locked trophy are already unlocked.

Megamouse commented:
@isshininu lol... Yeah I made a mistake. Should work in the next build

GitArUs commented:
Tested on unnamed game. Locking selected trophy by double click worked well. I have unlocked it again by performing appropriate action in game, but platinum remains locked.

Megamouse commented:
Can you guys please upload logs when after testing. I can't really deduce anything with that information alone.

GitArUs commented:
Here You are. You will see two unlocks there, the second one failed, because already_unlocked.
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/3980336/RPCS3.log.gz)

Megamouse commented:
can you please upload the trophy folder "dev_hdd0\home\00000001\trophy\NPWR11866_00"

GitArUs commented:
I hope You don't need all png icons ?...
[tropdir.tar.gz](https://github.com/RPCS3/rpcs3/files/3980386/tropdir.tar.gz)

isshininu commented:
Here is my log after reverting one of the bronze trophys to locked state and trying to achieve it again. Bronze achieved, platinum is not.
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/3989699/RPCS3.log.gz)

Megamouse commented:
Okay, let's try this by unlocking it ourself, since the game obviously didn't care to do it.

isshininu commented:
Unlocking platinum?

Megamouse commented:
no. I mean just try it again as before.

isshininu commented:
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/3990018/RPCS3.log.gz)
notning

Megamouse commented:
please also send the trophies again

GitArUs commented:
Doesn't work for me either.
[RPCS3_2.log.gz](https://github.com/RPCS3/rpcs3/files/3990606/RPCS3_2.log.gz)
[tropdir_2.tar.gz](https://github.com/RPCS3/rpcs3/files/3990608/tropdir_2.tar.gz)

Megamouse commented:
Allright. I noticed that a specific value was missing because old trophy files do not contain it yet.
This should now work even with old data.

I also noticed that I was restricting the unlock mechanism to a certain condition, which was illogical (but didn't cause the problem you guys are having).

isshininu commented:
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/3990988/RPCS3.log.gz)
Yeah,  it works. No platinum notification in game, but after unlocking bronze platinum WAS here in trophy manager..

GitArUs commented:
Confirmed. Platinum unlocked, but no trophy popup (maybe hidden under first popup ?).
[RPCS3_3.log.gz](https://github.com/RPCS3/rpcs3/files/3991202/RPCS3_3.log.gz)
[tropdir_3.tar.gz](https://github.com/RPCS3/rpcs3/files/3991203/tropdir_3.tar.gz)

Megamouse commented:
Yes, showing the trophies queued or at the same time isn't implemented yet

kd-11 commented:
Queued popups implemented with https://github.com/RPCS3/rpcs3/pull/7186. If building locally, you can check if combining the 2 PRs makes popups work, otherwise just wait a bit for it to be merged.

GitArUs commented:
I'm not sure if I did it right (merged one commit from #7184 and two commits from #7186 to this PR). Popup animations work (slide-in and slide-out), but still only one trophy popup is visible, platinum trophy unlocks but there is no platinum popup.

Megamouse commented:
I haven't applied the new logic yet, so there's no way it could've worked yet.

Megamouse commented:
You can try again now.

Megamouse commented:
@isshininu can you please test again?

isshininu commented:
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/4023713/RPCS3.log.gz)
![Снимок экрана (1012)](https://user-images.githubusercontent.com/6324523/71783474-bac7a200-3000-11ea-9bdb-e361602ef95f.png)
![Снимок экрана (1013)](https://user-images.githubusercontent.com/6324523/71783475-bac7a200-3000-11ea-8805-54dfc24c68d3.png)

kd-11 commented:
Nice

GitArUs commented:
That was fast. Thanks. Works as expected.
